File: app/src/main/java/agersant/polaris/features/player/PlayerFragment.java
package agersant.polaris.features.player;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.SeekBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import agersant.polaris.App;
import agersant.polaris.CollectionItem;
import agersant.polaris.PlaybackQueue;
import agersant.polaris.PolarisPlayer;
import agersant.polaris.PolarisState;
import agersant.polaris.R;
import agersant.polaris.api.API;
import agersant.polaris.databinding.FragmentPlayerBinding;

public class PlayerFragment extends Fragment {

    private FragmentPlayerBinding binding;
    private boolean seeking = false;
    private BroadcastReceiver receiver;
    private Handler seekBarUpdateHandler;
    private Runnable updateSeekBar;
    private API api;
    private PolarisPlayer player;
    private PlaybackQueue playbackQueue;

    private void subscribeToEvents() {
        final PlayerFragment that = this;
        IntentFilter filter = new IntentFilter();
        filter.addAction(PolarisPlayer.PLAYING_TRACK);
        filter.addAction(PolarisPlayer.PAUSED_TRACK);
        filter.addAction(PolarisPlayer.RESUMED_TRACK);
        filter.addAction(PolarisPlayer.COMPLETED_TRACK);
        filter.addAction(PolarisPlayer.OPENING_TRACK);
        filter.addAction(PolarisPlayer.BUFFERING);
        filter.addAction(PolarisPlayer.NOT_BUFFERING);
        filter.addAction(PlaybackQueue.CHANGED_ORDERING);
        filter.addAction(PlaybackQueue.QUEUED_ITEM);
        filter.addAction(PlaybackQueue.QUEUED_ITEMS);
        filter.addAction(PlaybackQueue.REMOVED_ITEM);
        filter.addAction(PlaybackQueue.REMOVED_ITEMS);
        filter.addAction(PlaybackQueue.REORDERED_ITEMS);
        receiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                switch (intent.getAction()) {
                    case PolarisPlayer.OPENING_TRACK:
                    case PolarisPlayer.BUFFERING:
                    case PolarisPlayer.NOT_BUFFERING:
                        that.updateBuffering();
                    case PolarisPlayer.PLAYING_TRACK:
                        that.updateContent();
                        that.updateControls();
                        break;
                    case PolarisPlayer.PAUSED_TRACK:
                    case PolarisPlayer.RESUMED_TRACK:
                    case PolarisPlayer.COMPLETED_TRACK:
                    case PlaybackQueue.CHANGED_ORDERING:
                    case PlaybackQueue.REMOVED_ITEM:
                    case PlaybackQueue.REMOVED_ITEMS:
                    case PlaybackQueue.REORDERED_ITEMS:
                    case PlaybackQueue.QUEUED_ITEM:
                    case PlaybackQueue.QUEUED_ITEMS:
                    case PlaybackQueue.OVERWROTE_QUEUE:
                        that.updateControls();
                        break;
                }
            }
        };
        App.instance.registerReceiver(receiver, filter);
    }

    private void scheduleSeekBarUpdates() {
        updateSeekBar = () -> {
            if (!seeking) {
                int precision = 10000;
                float position = player.getPositionRelative();
                binding.seekBar.setMax(precision);
                binding.seekBar.setProgress((int) (precision * position));
            }
            seekBarUpdateHandler.postDelayed(updateSeekBar, 20/*ms*/);
        };
        seekBarUpdateHandler.post(updateSeekBar);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        PolarisState state = App.state;
        api = state.api;
        player = state.player;
        playbackQueue = state.playbackQueue;
        seekBarUpdateHandler = new Handler();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentPlayerBinding.inflate(inflater);

        binding.seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            int newPosition = 0;

            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                newPosition = progress;
            }

            public void onStartTrackingTouch(SeekBar seekBar) {
                seeking = true;
            }

            public void onStopTrackingTouch(SeekBar seekBar) {
                player.seekToRelative((float) newPosition / seekBar.getMax());
                seeking = false;
                updateControls();
            }
        });

        binding.skipNext.setOnClickListener((View) -> {
            player.skipNext();
        });
        binding.skipPrevious.setOnClickListener((View) -> {
            player.skipPrevious();
        });
        binding.pauseToggle.setOnClickListener((View) -> {
            if (player.isPlaying()) {
                player.pause();
            } else {
                player.resume();
            }
        });

        refresh();

        return binding.getRoot();
    }

    private void refresh() {
        updateContent();
        updateControls();
        updateBuffering();
    }

    @Override
    public void onStart() {
        subscribeToEvents();
        scheduleSeekBarUpdates();
        super.onStart();
    }

    @Override
    public void onStop() {
        App.instance.unregisterReceiver(receiver);
        receiver = null;
        super.onStop();
    }

    @Override
    public void onResume() {
        super.onResume();
        refresh();
    }

    private void updateContent() {
        CollectionItem currentItem = player.getCurrentItem();
        if (currentItem != null) {
            populateWithTrack(currentItem);
        }
    }

    private void updateControls() {
        final float disabledAlpha = 0.2f;

        int playPauseIcon = player.isPlaying() ? R.drawable.ic_pause_black_24dp : R.drawable.ic_play_arrow_black_24dp;
        binding.pauseToggle.setImageResource(playPauseIcon);
        binding.pauseToggle.setAlpha(player.isIdle() ? disabledAlpha : 1.f);

        if (playbackQueue.hasNextTrack(player.getCurrentItem())) {
            binding.skipNext.setClickable(true);
            binding.skipNext.setAlpha(1.0f);
        } else {
            binding.skipNext.setClickable(false);
            binding.skipNext.setAlpha(disabledAlpha);
        }

        if (playbackQueue.hasPreviousTrack(player.getCurrentItem())) {
            binding.skipPrevious.setClickable(true);
            binding.skipPrevious.setAlpha(1.0f);
        } else {
            binding.skipPrevious.setClickable(false);
            binding.skipPrevious.setAlpha(disabledAlpha);
        }
    }

    private void updateBuffering() {
        if (player.isOpeningSong()) {
            binding.buffering.setText(R.string.player_opening);
        } else if (player.isBuffering()) {
            binding.buffering.setText(R.string.player_buffering);
        }
        if (player.isPlaying() && (player.isOpeningSong() || player.isBuffering())) {
            binding.buffering.setVisibility(View.VISIBLE);
        } else {
            binding.buffering.setVisibility(View.INVISIBLE);
        }

    }

    private void populateWithTrack(CollectionItem item) {
        assert item != null;

        String title = item.getTitle();
        if (title != null) {
            binding.title.setText(title);
        }

        String artist = item.getArtist();
        if (artist != null) {
            binding.artist.setText(artist);
        }

        String album = item.getAlbum();
        if (album != null) {
            binding.album.setText(album);
        }

        String artworkPath = item.getArtwork();
        if (artworkPath != null) {
            api.loadImageIntoView(item, binding.artwork);
        }
    }
}

File: app/src/main/java/agersant/polaris/features/player/PlayerFragment.kt
package agersant.polaris.features.player

import agersant.polaris.*
import agersant.polaris.api.API
import agersant.polaris.databinding.FragmentPlayerBinding
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.math.MathUtils
import androidx.fragment.app.Fragment
import com.google.android.material.slider.Slider
import kotlin.math.roundToInt

class PlayerFragment : Fragment() {
    private var receiver: BroadcastReceiver? = null

    private lateinit var binding: FragmentPlayerBinding
    private lateinit var seekBarUpdateHandler: Handler
    private lateinit var updateSeekBar: Runnable
    private lateinit var api: API
    private lateinit var player: PolarisPlayer
    private lateinit var playbackQueue: PlaybackQueue
    private var seeking = false

    private fun subscribeToEvents() {
        val that = this
        val filter = IntentFilter()
        filter.addAction(PolarisPlayer.PLAYING_TRACK)
        filter.addAction(PolarisPlayer.PAUSED_TRACK)
        filter.addAction(PolarisPlayer.RESUMED_TRACK)
        filter.addAction(PolarisPlayer.COMPLETED_TRACK)
        filter.addAction(PolarisPlayer.OPENING_TRACK)
        filter.addAction(PolarisPlayer.BUFFERING)
        filter.addAction(PolarisPlayer.NOT_BUFFERING)
        receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                when (intent.action) {
                    PolarisPlayer.OPENING_TRACK, PolarisPlayer.BUFFERING, PolarisPlayer.NOT_BUFFERING -> {
                        that.updateBuffering()
                        that.updateContent()
                        that.updateControls()
                    }
                    PolarisPlayer.PLAYING_TRACK -> {
                        that.updateContent()
                        that.updateControls()
                    }
                    PolarisPlayer.PAUSED_TRACK,
                    PolarisPlayer.RESUMED_TRACK,
                    PolarisPlayer.COMPLETED_TRACK,
                    -> that.updateControls()
                }
            }
        }
        App.instance.registerReceiver(receiver, filter)
        playbackQueue.liveItems.observe(viewLifecycleOwner) { updateControls() }
        playbackQueue.liveOrdering.observe(viewLifecycleOwner) { updateControls() }
    }

    private fun scheduleSeekBarUpdates() {
        updateSeekBar = Runnable {
            val duration = player.duration / 1000
            val position = player.positionRelative
            updateTime(position, duration)
            seekBarUpdateHandler.postDelayed(updateSeekBar, 20 /*ms*/)
        }
        seekBarUpdateHandler.post(updateSeekBar)
    }

    private fun updateTime(position: Float, duration: Float) {
        val clamped = MathUtils.clamp(position, 0f, 1f)
        if (!seeking) binding.controls.slider.value = clamped
        binding.controls.currentTime.text = formatTime(clamped * duration)
        binding.controls.totalTime.text = formatTime(duration)
    }

    private fun formatTime(time: Float): String {
        val timeSeconds = if (time.isNaN()) 1 else time.roundToInt()
        val minutes = timeSeconds / 60
        val seconds = String.format("%02d", timeSeconds % 60)
        return "$minutes:$seconds"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val state = App.state
        api = state.api
        player = state.player
        playbackQueue = state.playbackQueue
        seekBarUpdateHandler = Handler()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?, savedInstanceState: Bundle?,
    ): View {
        binding = FragmentPlayerBinding.inflate(inflater)
        binding.controls.slider.setLabelFormatter { formatTime(it * player.duration / 1000) }
        binding.controls.slider.addOnSliderTouchListener(object : Slider.OnSliderTouchListener {
            override fun onStartTrackingTouch(slider: Slider) {
                seeking = true
            }

            override fun onStopTrackingTouch(slider: Slider) {
                player.seekToRelative(slider.value / slider.valueTo)
                seeking = false
                updateControls()
            }
        })
        binding.controls.skipNext.setOnClickListener { player.skipNext() }
        binding.controls.skipPrevious.setOnClickListener { player.skipPrevious() }
        binding.controls.pauseToggle.setOnClickListener {
            if (player.isPlaying) {
                player.pause()
            } else {
                player.resume()
            }
        }
        refresh()
        return binding.root
    }

    private fun refresh() {
        updateContent()
        updateControls()
        updateBuffering()
    }

    override fun onStart() {
        subscribeToEvents()
        scheduleSeekBarUpdates()
        super.onStart()
    }

    override fun onStop() {
        App.instance.unregisterReceiver(receiver)
        receiver = null
        super.onStop()
    }

    override fun onResume() {
        super.onResume()
        refresh()
    }

    private fun updateContent() {
        populateWithTrack(player.currentItem)
    }

    private fun updateControls() {
        val disabledAlpha = 0.2f
        val playPauseIcon = if (player.isPlaying) R.drawable.baseline_pause_24 else R.drawable.baseline_play_arrow_24
        binding.controls.pauseToggle.setImageResource(playPauseIcon)
        binding.controls.pauseToggle.alpha = if (player.isIdle) disabledAlpha else 1f
        if (playbackQueue.hasNextTrack(player.currentItem)) {
            binding.controls.skipNext.isClickable = true
            binding.controls.skipNext.alpha = 1.0f
        } else {
            binding.controls.skipNext.isClickable = false
            binding.controls.skipNext.alpha = disabledAlpha
        }
        if (playbackQueue.hasPreviousTrack(player.currentItem)) {
            binding.controls.skipPrevious.isClickable = true
            binding.controls.skipPrevious.alpha = 1.0f
        } else {
            binding.controls.skipPrevious.isClickable = false
            binding.controls.skipPrevious.alpha = disabledAlpha
        }
    }

    private fun updateBuffering() {
        if (player.isPlaying && (player.isOpeningSong || player.isBuffering)) {
            binding.controls.loading.show()
        } else {
            binding.controls.loading.hide()
        }

        binding.controls.loading
    }

    private fun populateWithTrack(item: CollectionItem?) {
        binding.controls.title.text = item?.title ?: resources.getString(R.string.player_no_song)
        binding.controls.artist.text = item?.artist ?: resources.getString(R.string.player_unknown)
        binding.controls.album.text = item?.album ?: resources.getString(R.string.player_unknown)

        if (item?.artwork != null)
            api.loadImageIntoView(item, binding.artwork)
        else
            binding.artwork.setImageResource(R.drawable.launcher_icon_foreground)
    }
}